### PR TITLE
Create C# object for BaseDevice

### LINKED ISSUE #110: Create C# object for BaseDevice
#80 

We need a C# object for a BaseDevice. This should allow us to convert data into a BsonDocument.

### PR DESCRIPTION
**Pull request by roguedeleter**
Closes #110 

This PR implements a BaseDevice object that we can use to convert raw Device data into a BsonDocument. This is part of the "Factory" story #80, but I decided against implementing a Factory class. This type of work is easily done using a C# object instead, since there exist methods out there that automatically serialize objects into BSON.

### CODE REVIEW

File: BEIMA.DB.Schemas/device-schema.txt [hunk 1]
```diff
@@ -1,6 +1,5 @@
 {
     "deviceTypeId": String,
-    "serialNumber": String,
```
Comment (josephmoore99) [inline BEIMA.DB.Schemas/device-schema.txt:3]: Why are we removing this?

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        /// Full constructor to instantiate a Device object.
+        /// </summary>
+        /// <param name="id">ObjectId of the object itself.</param>
+        /// <param name="deviceTypeId">ObjectId of the DeviceType this Device is linked to.</param>
+        /// <param name="deviceTag"></param>
+        /// <param name="manufacturer"></param>
+        /// <param name="modelNum"></param>
+        /// <param name="serialNum"></param>
+        /// <param name="yearManufactured"></param>
+        /// <param name="notes"></param>
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:34]: params

File: BEIMA.DB.Schemas/device-schema.txt (hunk 1 above)
Comment (roguedeleter) [reply BEIMA.DB.Schemas/device-schema.txt:3]: "serialNumber" is a duplicate value, as "serialNum" already exists.

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 2 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs:34]: I thought these were self explanatory, what type of param comments need to be put here?

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        /// <summary>
+        /// Gets a BsonDocument that represents the current state of the BaseDevice object.
+        /// </summary>
+        /// <returns>BsonDocument that represents the current state of the BaseDevice object. Conforms to the schema located in BEIMA.DB.Schemas.</returns>
+        /// <exception cref="ArgumentNullException">Throws exception when any of the required fields are null.</exception>
+        public BsonDocument GetBsonDocument()
+        {
+            if (DeviceTag == null)
+            {
+                throw new ArgumentNullException($"BaseDevice - {nameof(DeviceTag)} is null");
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:95]: Not sure how I feel about throwing an exception here on null data, especially for things like Notes or LastModified that may not necessarily be required have any data in them. If we want to throw an exception, I think it would be better as a NullReferenceException instead of ArgumentNullException?

Comment (AshlynAdamson) [inline BEIMA.Backend/MongoService/BaseDevice.cs:95]: I think we can put this null check into it's own method, and pass in the field as the argument. That way we have multiple method calls, rather than almost duplicate if and throw statements

Comment (josephmoore99) [reply BEIMA.Backend/MongoService/BaseDevice.cs:95]: I think we should only throw exceptions on data we think should absolutely be required.

Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs:95]: Null data indicates that the programmer did not pass anything in for that field. The programmer needs to pass in an empty string, or some other default value. If null data has been passed in, it means something has gone wrong and an exception must be thrown.

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            }
+            return this.ToBsonDocument();
+        }
+
+        /// <summary>
+        /// Adds a custom field to the "fields" object.
+        /// </summary>
+        /// <param name="key">The ObjectId of the linked field in the linked DeviceType.</param>
+        /// <param name="value">The actual value of the field.</param>
+        public void AddField(string key, dynamic value)
```
Comment (AshlynAdamson) [inline BEIMA.Backend/MongoService/BaseDevice.cs:133]: This method adds a value for a custom field that was defined in the device type. Do we need a method to do that, but for a custom field on a single device? I think that was a scenario that Brian had described. Maybe this will be tackled in a future task, but thought I'd bring it up to make sure

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 3 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs:95]: Comparing the descriptions of both exceptions, it seemed like the wording of ArgumentNullException seems more appropriate to me.
https://docs.microsoft.com/en-us/dotnet/api/system.nullreferenceexception?view=net-6.0
https://docs.microsoft.com/en-us/dotnet/api/system.argumentnullexception?view=net-6.0

It seems like a NullReferenceException is thrown when "there is an attempt to dereference a null object reference."

Versus, a ArgumentNullException is thrown when "a null reference is passed to a method that does not accept it as a valid argument." This seems more appropriate, since the only way those values can be null, is if a null was passed in as an argument to the constructor/object initializer.

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        /// </summary>
+        /// <param name="id">ObjectId of the object itself.</param>
+        /// <param name="deviceTypeId">ObjectId of the DeviceType this Device is linked to.</param>
+        /// <param name="deviceTag"></param>
+        /// <param name="manufacturer"></param>
+        /// <param name="modelNum"></param>
+        /// <param name="serialNum"></param>
+        /// <param name="yearManufactured"></param>
+        /// <param name="notes"></param>
+        public BaseDevice(ObjectId id, ObjectId deviceTypeId, string deviceTag, string manufacturer, string modelNum, string serialNum, int yearManufactured, string notes)
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:35]:
```suggestion
        public BaseDevice(ObjectId id, ObjectId deviceTypeId, int yearManufactured, string deviceTag = string.Empty, string manufacturer = string.Empty, string modelNum = string.Empty, string serialNum = string.Empty, string notes = string.Empty)
```
We can set defaults for constructor parameters, also should probably split this into a couple lines.

**Review by AshlynAdamson — CHANGES_REQUESTED** [review]
Left some feedback in other comments. Looks great!

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        {
+            //BsonClassMap.RegisterClassMap<BaseDevice>();
+
+            Id = id;
+            DeviceTypeId = deviceTypeId;
+            DeviceTag = deviceTag ?? string.Empty;
+            Manufacturer = manufacturer ?? string.Empty;
+            ModelNum = modelNum ?? string.Empty;
+            SerialNum = serialNum ?? string.Empty;
+            YearManufactured = yearManufactured;
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:45]: Should we make this a nullable int? Doesn't seem like we should be requiring every device to have this instantiated.

File: BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            Assert.That(device.Id, Is.EqualTo(validObjId));
+            Assert.That(device.DeviceTypeId, Is.EqualTo(validDeviceTypeId));
+            Assert.That(device.DeviceTag, Is.EqualTo(validDeviceTag));
+            Assert.That(device.Manufacturer, Is.EqualTo(validManufacturer));
+            Assert.That(device.ModelNum, Is.EqualTo(validModelNum));
+            Assert.That(device.SerialNum, Is.EqualTo(validSerialNum));
+            Assert.That(device.YearManufactured, Is.EqualTo(validYearManufactured));
+            Assert.That(device.Notes, Is.EqualTo(validNotes));
+
+            device.SetLastModified(validDate, validUser);
```
Comment (josephmoore99) [inline BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs:48]: I think it's good practice to set all of the object stuff at the beginning, and than do all the assertions at the end.

File: BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            device.AddField(validKey2, validValue2);
+            device.AddField(validKey3, validValue3);
+            var fields = device.Fields;
+            Assert.That((string)fields.GetElement(validKey1).Value, Is.EqualTo(validValue1));
+            Assert.That((bool)fields.GetElement(validKey2).Value, Is.EqualTo(validValue2));
+            Assert.That((int)fields.GetElement(validKey3).Value, Is.EqualTo(validValue3));
+        }
+
+        [Test]
+        public void BaseDeviceNotInstantiated_InstantiateUsingObjectInitializerWithAllNullStringValues_GetBsonDocumentThrowsException()
```
Comment (josephmoore99) [inline BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs:114]: If you want, you can set up various test cases using the same method. You would use something like:
`[TestCase(1, "2", null)]
[TestCase(3, "4", "5"]
public void MyTest(int number, string name, string value)`
Although, this would probably need a ridiculous number of method parameters .

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            Notes = notes ?? string.Empty;
+            Fields = new BsonDocument();
+            Location = new BsonDocument();
+            LastModified = new BsonDocument();
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:49]: We can just call the other constructor for these.

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+            Fields.Set(key, value);
+        }
+
+        /// <summary>
+        /// Sets all of the fields in the "location" object of the BaseDevice.
+        /// </summary>
+        /// <param name="buildingId"></param>
+        /// <param name="notes"></param>
+        /// <param name="latitude"></param>
+        /// <param name="longitude"></param>
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:144]: params

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        public void SetLocation(ObjectId buildingId, string notes, string latitude, string longitude)
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:145]:
```suggestion
        public void SetLocation(ObjectId buildingId, string notes = string.Empty, string latitude = string.Emtpy, string longitude = string.Empty)
```

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+            Location.Set("latitude", latitude);
+            Location.Set("longitude", longitude);
+        }
+
+        /// <summary>
+        /// Sets all of the fields in the "lastModified" object of the BaseDevice.
+        /// </summary>
+        /// <param name="date">Date of when the Device was last modified.</param>
+        /// <param name="user">Username of the user who last modified the Device.</param>
+        public void SetLastModified(DateTime? date, string user)
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:163]:
```suggestion
        public void SetLastModified(DateTime? date, string user = string.Empty)
```

**Review by josephmoore99 — CHANGES_REQUESTED** [review]
Good stuff, will be nice to have some more well enforced typing for database objects.

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 5 above)
Comment (josephmoore99) [reply BEIMA.Backend/MongoService/BaseDevice.cs:35]: I think you may still be able to manually set the parameters to null when you call it if you specifically specify it.

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 2 above)
Comment (josephmoore99) [reply BEIMA.Backend/MongoService/BaseDevice.cs:34]: You can literally just copy the param name if it's simple enough. Like "The device tag" for "deviceTag".

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 4 above)
Comment (josephmoore99) [reply BEIMA.Backend/MongoService/BaseDevice.cs:133]: I think that for now, this is fine, but in the future, I think we would likely want to use some sort of interaction with a deviceType object when setting these fields. Or even just query the database to see of the field value is valid for the given device type.

File: BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs (hunk 7 above)
Comment (Kenny-Miller) [reply BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs:48]: Would we want to just pull the assertions at the top out into their own test case?

Comment (josephmoore99) [reply BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs:48]: We shouldn't need to, setting the various properties should not be affecting one another.

Comment (AshlynAdamson) [reply BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs:48]: All the fields need to be set for a Bson to be returned correctly. There's another test case that only sets a few fields and leaves the rest null. Moving around the asserts is a good idea.

**Review by Kenny-Miller — APPROVED** [review]

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 2 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: Resolved [95e7484](https://github.com/cs481-ekh/s22-beima/pull/114/commits/95e7484f1c1e7fa2a0b2e1fa32c92df232cf3a0d)

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+        /// <param name="deviceTypeId">ObjectId of the DeviceType this Device is linked to.</param>
+        /// <param name="deviceTag"></param>
+        /// <param name="manufacturer"></param>
+        /// <param name="modelNum"></param>
+        /// <param name="serialNum"></param>
+        /// <param name="yearManufactured"></param>
+        /// <param name="notes"></param>
+        public BaseDevice(ObjectId id, ObjectId deviceTypeId, string deviceTag, string manufacturer, string modelNum, string serialNum, int yearManufactured, string notes)
+        {
+            //BsonClassMap.RegisterClassMap<BaseDevice>();
```
[the remark this answers is not in the document]
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: Resolved [95e7484](https://github.com/cs481-ekh/s22-beima/pull/114/commits/95e7484f1c1e7fa2a0b2e1fa32c92df232cf3a0d)

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+            YearManufactured = yearManufactured;
+            Notes = notes ?? string.Empty;
+            Fields = new BsonDocument();
+            Location = new BsonDocument();
+            LastModified = new BsonDocument();
+        }
+
+
+        [BsonId]
+        public ObjectId Id { get; set; }
```
[the remark this answers is not in the document]
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs:15]: Fixed [6d85cb2](https://github.com/cs481-ekh/s22-beima/pull/114/commits/6d85cb251c5a87d0b5701a4653d6c9306c76de53)

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 5 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: @josephmoore99 @AshlynAdamson If the programmer can still technically pass in null (either unintentionally or intentionally), then setting defaults for constructor parameters is not a good fix. The ??= syntax I've used in this constructor will take care of both defaults and null values.

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 6 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: I'm not sure if MongoDB allows nullable ints, I will look into this.

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 5 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: The way I envisioned the full constructor being used, is that the programmer will set variables like so:
`
var id = ...
var deviceTypeId = ...
var deviceTag = ...
(and then all the other properties) 
`
Then, they will pass in all those parameters into the constructor:
`
var device = new BaseDevice(id, deviceTypeId, deviceTag, manufacturer, modelNum, serialNum, yearManufactured, notes);
`

What if in the first block of code, one of those parameters was set to null (because maybe the front end has a bug that made it send a null)? Then, my current implementation will handle this properly and set a default value.

Comment (josephmoore99) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: I think it's nice to have though because it means you aren't forced to specify each parameter every single time, especially if you don't have the associated values.

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 6 above)
Comment (josephmoore99) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: If anything, we could have it be set to -1 or something.

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 5 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: @josephmoore99 The programmer must specify each parameter each time. There should not be a case where we only specify some fields and not all. The endpoints we write will be generalized, and we must assume that all of the fields will be populated with some sort of data. If it turns out that one field is null, it is okay since the constructor will handle it properly.

Comment (josephmoore99) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: Alright, if that's the case we can leave it as is. It was more of a quality of life improvement.

Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: I definitely understand the point about quality of life, since typing all of those parameters is such a pain. However, in this case, we would have to sacrifice some code/type safety which I am not very comfortable with on such a critical object.

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 6 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: MongoDB can store null values, but there is no way to truly specify it as an int if it is null. I am just going to store -1 as the default value as an alternative.

File: BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs (hunk 7 above)
Comment (roguedeleter) [reply BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs]: Resolved [bbc776a](https://github.com/cs481-ekh/s22-beima/pull/114/commits/bbc776a0b050b6457993e35109a7704be0e6b827)

File: BEIMA.Backend/MongoService/BaseDevice.cs (hunk 6 above)
Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: Resolved [401e45f](https://github.com/cs481-ekh/s22-beima/pull/114/commits/401e45f2842b9ab766d956b9ff5b098c2098b2a0)

File: BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs (hunk 8 above)
Comment (roguedeleter) [reply BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs:116]: Since I've already written it this way, and also the number of method parameters is so large, I will keep it like this. Going forward, I will try and make use of TestCases.

**Comment by roguedeleter** [comment]
All open issues have been resolved

**Review by AshlynAdamson — APPROVED** [review]

File: BEIMA.Backend/MongoService/BaseDevice.cs [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+            Location = new BsonDocument();
+            LastModified = new BsonDocument();
+        }
+
+        /// <summary>
+        /// Internal method to check if a particular property of this class is null
+        /// </summary>
+        /// <param name="arg">The Property being null checked.</param>
+        /// <param name="name">The name of the Property, to be returned in the exception message.</param>
+        /// <exception cref="ArgumentNullException"></exception>
```
Comment (josephmoore99) [inline BEIMA.Backend/MongoService/BaseDevice.cs:88]: Forgot exception description.

Comment (roguedeleter) [reply BEIMA.Backend/MongoService/BaseDevice.cs]: Resolved [7c9c90e](https://github.com/cs481-ekh/s22-beima/pull/114/commits/7c9c90e00b2aae595377c7263457998fac01415b)

**Review by josephmoore99 — APPROVED** [review]

**Review by AshlynAdamson — APPROVED** [review]
